### DISCUSSION TITLE
StackStorm Performance Benchmarking

### DISCUSSION THREAD

**Discussion by Bindu-yadav8** [StackStorm/st2#5510] [Q&A]
Hello Everyone,

We are using  StackStorm version st2 3.4.1, on Python 3.6.9 and using Orquesta workflows. Required help on the performance benchmarking performed on StackStorm. 

Please share any snippets, links and some images. 

If anybody has performed benchmark testing on StackStorm and deployed on the production environment. What were the challenges faced and what type of servers were used to achieve it.

@armab Could you please provide some insights on this?

Thanks!

**Comment by arm4b** [comment]
We've seen performance benchmarks performed by @userlocalhost for st2 v3.5.0 here: https://www.slideshare.net/Hiroyasu-OHYAMA/verification-of-performance-improvement-of-stackstorm-v35
Perhaps he could share the workflows that were used.

**Comment by Bindu-yadav8** [comment]
Hi @userlocalhost, Could you please share the workflows that were used to perform Benchmark Testing?

Also, please share what were the challenges faced type of servers used.

Thanks!

**Reply by userlocalhost** [reply]
@armab 
Thank you very much for introducing my slide!

@Bindu-yadav8 
Sure, I checked st2 performance using following action.

* [meetup_jp.test_concurrency](https://github.com/userlocalhost/st2_meetup_jp_demo/blob/master/actions/test_concurrency.yaml)
  - run many (small) nested tasks (This is renamed from `meetup_jp.test_items` of [this slide](https://www.slideshare.net/Hiroyasu-OHYAMA/verification-of-performance-improvement-of-stackstorm-v35))

* [meetup_jp.test_passing_data](https://github.com/userlocalhost/st2_meetup_jp_demo/blob/master/actions/test_passing_data.yaml)
  - pass large number of small data to subsequent task

(c.f. https://github.com/userlocalhost/st2_meetup_jp_demo)

And we have been faced with the problem that st2 takes time for map/reduce processing with large number of data. 
We run an operation to pick-up servers that meets specific criterion from our production environment. In order to this, we made a workflow to scan each server and check its condition, then run another task with picked-up server. But it takes time (than we expected) when st2 handle large number of servers at once. When this workflow targeted 1 server, it finished a couple of seconds on average. However, once it targeted 1,000 servers, it took far more than 1,000 times.
That's the reason why I'm interested in st2 performance.

**Comment by Bindu-yadav8** [comment]
We would like to know Performance of Stackstorm on Kubernetes deployment. The Question here is that how many events it can process per second and also want to know Server Configurations . It could be a single server, or on cluster.

Also, would like to understand if the Stackstorm is Scalable? If yes, could you please share more details on that.

Appreciate your response at the earliest.

@armab, @userlocalhost

Thanks!

**Reply by arm4b** [reply]
You can find the StackStorm HA/Scalability doc at https://docs.stackstorm.com/reference/ha.html
